lib/Transforms/ShapeInference/ShapeInference.cpp

```diff
@@ -0,0 +1,187 @@
+
+    // Check if there are even argument attributes at all.
+    // If not, there is nothing to infer from -> exit
+    auto optionalAttrs = funcOp.getArgAttrs();
+    if (!optionalAttrs) return failure();
+    auto attrs = optionalAttrs->getValue();
+
+    // Update types with annotations (where it exists)
+    bool changed = false;
+    for (auto &[idx, type] : dynTensorArgs) {
+      // Check if the current argument has any attributes.
+      auto dictattr = dyn_cast<DictionaryAttr>(attrs[idx]);
+      if (!dictattr) continue;  // nothing to infer from
+
+      // Get the shape attribute if it exists
+      auto *shapeAttr = llvm::find_if(dictattr, [&](NamedAttribute attr) {
+        return attr.getName() == "shape.shape";
+      });
+      if (!shapeAttr) continue;  // nothing to infer from
+
+      // Extract the actual shape from the shape annotation attribute
+      auto arrayAttr = dyn_cast<ArrayAttr>(shapeAttr->getValue());
+      if (!arrayAttr)
+        return emitError(funcOp.getLoc(),
+                         "Invalid shape attribute (expected ArrayAttr)");
+
+      if (llvm::any_of(arrayAttr,
+                       [](Attribute attr) { return !isa<IntegerAttr>(attr); }))
+        return emitError(
+            funcOp.getLoc(),
+            "Invalid shape attribute (expected ArrayAttr of only IntegerAttr)");
+
+      SmallVector<int64_t> shape = llvm::to_vector(
+          llvm::map_range(arrayAttr.getAsRange<IntegerAttr>(),
+                          [](IntegerAttr i) { return i.getInt(); }));
+
+      if (shape.size() != type.getRank())
+        return emitError(
+            funcOp.getLoc(),
+            llvm::formatv("Shape annotation has wrong number of elements "
+                          "(expected {0}, got {1}).",
+                          type.getRank(), shape.size()));
+
+      // save the updated argument type
+      type = RankedTensorType::get(shape, type.getElementType());
+      changed = true;
+    }
+    // no new information found to infer from -> exit
+    if (!changed) return failure();
+
+    rewriter.startOpModification(funcOp);
+
+    /////////////////////////////////////////////////////////////////////
+    ////  Step 2: Update Function Type/Signature based on Attributes ////
+    /////////////////////////////////////////////////////////////////////
+    // We need to update the function type to reflect the new types,
+    // update the type of the "real" arguments (block arguments),
+    // and we also remove the attribute argument from the signature
+    // (primarily for better readability of the IR after shape inference).
+    //
+    // We could delay updating the function type until the end,
+    // when we also know the inferred return type(s),
+    // but that would leave the func op in a less consistent state longer
+    auto newInputTypes = llvm::to_vector(funcOp.getFunctionType().getInputs());
+    for (auto [idx, type] : dynTensorArgs) {
+      funcOp.getArgument(idx).setType(type);
```

Comment:
I'm assuming you didn't access the actual block arguments above to support function declarations with this pass as well (otherwise maybe you would want to use `AttributeUtils::findAttributeAssociatedWith` to simplify the lookups above). But if my assumption is correct, then this will fail if the function is just a declaration.

```suggestion
      if (!funcOp.isDeclaration()) funcOp.getArgument(idx).setType(type);
```

Reply:
Actually, I don't think I can support declarations of functions, because how would I determine what the static shape of a return type of, e.g., `tensor<?x?xi32>` should be, so I'll just guard the whole pattern against `isDeclaration()`. I *will* switch to `findAttributeAssociatedWith`, just didn't remember that it existed.

Reply:
(I guess we could extend the shape annotation to also work for return arguments (does `func.func` have attributes for its returns?) but I think that's out of scope for now)

Reply:
> That seems to be looking at attributes on the actual block args, though, where the ArgAttrs for a funcOp are just in a special dictionary inside the main op attribute dictionary for the func op itself?

That is handled as a special case for different op types, because block arguments can't actually have attributes: https://github.com/google/heir/blob/1b11fcf16add488f84cd57b7f107e4e2a046a551/lib/Utils/AttributeUtils.cpp#L34-L36

> does func.func have attributes for its returns?

Yes, `getResultAttr` works similarly to `getOperandAttr`. So function declarations should be supportable this way. Let's do it in a followup.